code/Controllers/CMSMain.php

```diff
@@ -1006,12 +1006,8 @@ public function save($data, $form) {
 
 		// Update the class instance if necessary
 		if(isset($data['ClassName']) && $data['ClassName'] != $record->ClassName) {
-			$newClassName = $record->ClassName;
```

Comment:
This is the part that was incorrectly creating the new instance with the wrong physical class.